### DISCUSSION TITLE
How do I chain several nullable where expressions?

### DISCUSSION THREAD

**Discussion by ajilo-eGov** [simolus3/drift#2266] [Q&A] [answered]
I have a dart object that I use to query. Something like this

```dart
class UserQueryModel {
  final String? id;
  final String? name;
  final int? age;

  const UserQueryModel({this.id, this.name, this.age});
}
```

My requirement is to run a `select` on the drift table and use the above model as a filter. i.e. filter the table with `id` **ONLY** if `id` is not-null, otherwise return all records, etc.

I have something like this right now

```dart
select(user)..where((table) {
  final idExp = query.id == null ? null : table.id.equals(query.id);
  final ageExp = query.age == null ? null : table.age.equals(query.age);

  // This is what I need to do
  return (idExp ?? Expression(true)) & (ageExp ?? Expression(true));
});
```

But the `Expression` object is abstract and I am not sure if there is a better way to do this.

**Comment by simolus3** [comment, marked as answer]
> `Expression(true)`

You can replace this with `Constant(true)` which would be the canonical way to embed a SQL literal.

In general, it's a good approach. You could also filter out the nullable expressions:

```dart
Expression<bool> toCondition(Iterable<Expression<bool>?> conditions) {
  final existingConditions = conditions.whereType<Expression<bool>>();
  if (existingConditions.isEmpty) return const Constant(true);

  return existingConditions.skip(1).fold(existingConditions.first, (chain, cond) => chain & cond);
}
```

**Reply by North101** [reply]
I have this util method which is similar to @simolus3 

```dart
Expression<bool> buildAnd(Iterable<Expression<bool>> iterable) {
  if (iterable.isEmpty) return const Constant(true);
  return iterable.reduce((value, element) => value & element);
}

Expression<bool> buildOr(Iterable<Expression<bool>> iterable) {
  if (iterable.isEmpty) return const Constant(true);
  return iterable.reduce((value, element) => value | element);
}
```

and if something is optional I will write it like:
```dart
final where = buildAnd([
  if (query.id != null) table.id.equals(query.id!),
  if (query.age != null) table.age.equals(query.age!),
]);
```

**Reply by simolus3** [reply]
Oh yeah, reduce is a much better choice that I keep forgetting about 😅

**Reply by ajilo-eGov** [reply]
Thanks @North101  and @simolus3. This logic makes sense. 

But FYI, the below block does not work.

```dart
Expression<bool> buildAnd(Iterable<Expression<bool>> iterable) {
  if (iterable.isEmpty) return const Constant(true);
  return iterable.reduce((value, element) => value & element);
}
```

It gives the following error
> The return type 'Expression<bool?>' isn't a 'Expression<bool>', as required by the closure's context

Not sure if it's something I am doing wrong. Made a temporary fix like below. Not sure if the logic is correct though.

```dart
Expression<bool> buildAnd(Iterable<Expression<bool?>> iterable) {
  if (iterable.isEmpty) return const Constant(true);
  final result =  iterable.reduce((value, element) => value & element);

  return result.equals(true);
}
```
